From: Danila Tikhonov <danila@jiaxyga.com>
Date: Thu, 30 Mar 2023 21:18:07 +0300
Subject: arm64: dts: qcom: sm7150: Add device-tree for Xiaomi Mi Note 10

Add device tree for the Xiaomi Mi Note 10 (Xiaomi Mi CC9 Pro) /
Xiaomi Mi Note 10 Pro (Xiaomi Mi CC9 Pro Premium Edition) (tucana)
smartphone. This device is based on Snapdragon 730G (sm7150) SoC.
---
 arch/arm64/boot/dts/qcom/Makefile                  |   1 +
 ...xiaomi-davinci.dts => sm7150-xiaomi-tucana.dts} | 347 ++++++++++++++-------
 2 files changed, 235 insertions(+), 113 deletions(-)
 copy arch/arm64/boot/dts/qcom/{sm7150-xiaomi-davinci.dts => sm7150-xiaomi-tucana.dts} (50%)

diff --git a/arch/arm64/boot/dts/qcom/Makefile b/arch/arm64/boot/dts/qcom/Makefile
index bf6c6a4..dff425a 100644
--- a/arch/arm64/boot/dts/qcom/Makefile
+++ b/arch/arm64/boot/dts/qcom/Makefile
@@ -277,6 +277,7 @@ dtb-$(CONFIG_ARCH_QCOM)	+= sm7150-xiaomi-davinci.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sm7150-xiaomi-surya-huaxing.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sm7150-xiaomi-surya-tianma.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sm7150-xiaomi-sweet.dtb
+dtb-$(CONFIG_ARCH_QCOM)	+= sm7150-xiaomi-tucana.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sm7225-fairphone-fp4.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sm7325-nothing-spacewar.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sm8150-hdk.dtb
diff --git a/arch/arm64/boot/dts/qcom/sm7150-xiaomi-davinci.dts b/arch/arm64/boot/dts/qcom/sm7150-xiaomi-tucana.dts
similarity index 50%
copy from arch/arm64/boot/dts/qcom/sm7150-xiaomi-davinci.dts
copy to arch/arm64/boot/dts/qcom/sm7150-xiaomi-tucana.dts
index 8901aa1..fd7f944 100644
--- a/arch/arm64/boot/dts/qcom/sm7150-xiaomi-davinci.dts
+++ b/arch/arm64/boot/dts/qcom/sm7150-xiaomi-tucana.dts
@@ -1,14 +1,13 @@
 // SPDX-License-Identifier: BSD-3-Clause
 /*
- * SM7150-AA Xiaomi Mi 9T / Redmi K20 specific device tree
- *
  * Copyright (c) 2024, Danila Tikhonov <danila@jiaxyga.com>
- * Copyright (c) 2024, Jens Reidel <adrian@travitia.xyz>
+ * Copyright (c) 2024-2025, Alexander Baransky <alexander.baranskiy@yandex.ru>
  */
 
 /dts-v1/;
 
 #include "sm7150-xiaomi-common.dtsi"
+#include <dt-bindings/input/ti-drv260x.h>
 
 /delete-node/ &wlan_msa_mem;
 /delete-node/ &npu_mem;
@@ -30,26 +29,21 @@ &tz_mem {
 
 / {
 	/*
-	 * Xiaomi Mi 9T
-	 * Xiaomi Redmi K20
+	 * Xiaomi Mi Note 10 (Xiaomi Mi CC9 Pro)
+	 * Xiaomi Mi Note 10 Pro (Xiaomi Mi CC9 Pro Premium Edition)
 	 */
-	model = "Xiaomi Mi 9T";
-	compatible = "xiaomi,davinci", "qcom,sm7150";
-
-	qcom,board-id = <40 0>;
-
-	vreg_ibb: regulator-ibb {
-		compatible = "regulator-fixed";
-		regulator-name = "ibb";
-		regulator-min-microvolt = <6000000>;
-		regulator-max-microvolt = <6000000>;
-	};
-
-	vreg_lab: regulator-lab {
-		compatible = "regulator-fixed";
-		regulator-name = "lab";
-		regulator-min-microvolt = <6000000>;
-		regulator-max-microvolt = <6000000>;
+	model = "Xiaomi Mi Note 10";
+	compatible = "xiaomi,tucana", "qcom,sm7150";
+
+	qcom,board-id = <42 0>;
+
+	chosen {
+		/*
+		 * Serial engine with the debug UART operates in SPI mode on tucana
+		 * So disable console and remove stdout-path
+		 */
+		/delete-property/ stdout-path;
+		bootargs = "console=null";
 	};
 
 	reserved-memory {
@@ -80,13 +74,70 @@ gpu_mem: gpu@97f15000 {
 
 		rmtfs_mem: rmtfs@f2e01000 {
 			compatible = "qcom,rmtfs-mem";
-			reg = <0 0xf2e01000 0 0x600000>;
+			reg = <0 0xf2e01000 0 0x200000>;
 			no-map;
 
 			qcom,client-id = <1>;
 			qcom,vmid = <QCOM_SCM_VMID_MSS_MSA>;
 		};
 	};
+
+	sound {
+		compatible = "qcom,sm8250-sndcard";
+		pinctrl-0 = <&pri_mi2s_active &pri_mi2s_ws_active>;
+		pinctrl-names = "default";
+		model = "Xiaomi Mi Note 10";
+
+		mm1-dai-link {
+			link-name = "MultiMedia1";
+			cpu {
+				sound-dai = <&q6asmdai MSM_FRONTEND_DAI_MULTIMEDIA1>;
+			};
+		};
+
+		speaker_playback_dai {
+			link-name = "Primary Spkr Playback";
+			cpu {
+				sound-dai = <&q6afedai PRIMARY_MI2S_RX>;
+			};
+
+			platform {
+				sound-dai = <&q6routing>;
+			};
+
+			codec {
+				sound-dai = <&tfa9874>;
+			};
+		};
+	};
+
+	vreg_ibb: regulator-ibb {
+		compatible = "regulator-fixed";
+		regulator-name = "ibb";
+		regulator-min-microvolt = <6000000>;
+		regulator-max-microvolt = <6000000>;
+	};
+
+	vreg_lab: regulator-lab {
+		compatible = "regulator-fixed";
+		regulator-name = "lab";
+		regulator-min-microvolt = <6000000>;
+		regulator-max-microvolt = <6000000>;
+	};
+
+	vreg_touch_vdd: regulator-touch-vdd {
+		compatible = "regulator-fixed";
+		regulator-name = "touch_vdd";
+		gpio = <&tlmm 72 GPIO_ACTIVE_HIGH>;
+		startup-delay-us = <4000>;
+		enable-active-high;
+		regulator-boot-on;
+		vin-supply = <&vreg_l10a_1p8>;
+	};
+};
+
+&battery {
+	charge-full-design-microamp-hours = <5260000>;
 };
 
 &framebuffer {
@@ -96,7 +147,7 @@ &framebuffer {
 
 &gpu {
 	zap-shader {
-		firmware-name = "qcom/sm7150/xiaomi/davinci/a615_zap.mbn";
+		firmware-name = "qcom/sm7150/xiaomi/tucana/a615_zap.mbn";
 	};
 };
 
@@ -104,34 +155,70 @@ &i2c4 {
 	clock-frequency = <100000>;
 	status = "okay";
 
-	/* nxp,tfa9874 (speaker amplifier) @ 34 */
+	haptics@5a {
+		compatible = "ti,drv2605l";
+		reg = <0x5a>;
+		enable-gpios = <&tlmm 4 GPIO_ACTIVE_HIGH>;
+		mode = <DRV260X_LRA_NO_CAL_MODE>;
+		library-sel = <DRV260X_LIB_LRA>;
+		vib-rated-mv = <3200>;
+		vib-overdrive-mv = <3200>;
+	};
 };
 
 &i2c7 {
 	clock-frequency = <100000>;
 	status = "okay";
 
-	touchscreen: goodix@5d {
-		compatible = "goodix,gt9886";
-		reg = <0x5d>;
-		interrupts-extended = <&tlmm 9 IRQ_TYPE_LEVEL_LOW>;
-		reset-gpios = <&tlmm 8 GPIO_ACTIVE_LOW>;
+	touchscreen: fts@49 {
+		compatible = "st,fts";
+		reg = <0x49>;
+
+		interrupt-parent = <&tlmm>;
+		interrupts = <9 IRQ_TYPE_LEVEL_LOW>;
+
+		pinctrl-0 = <&ts_int_active &ts_reset_active>;
+		pinctrl-1 = <&ts_int_suspend &ts_reset_suspend>;
+		pinctrl-names = "default", "sleep";
+
+		vdd-supply = <&vreg_touch_vdd>;
 		avdd-supply = <&vreg_l6c_3p0>;
-		touchscreen-size-x = <1080>;
-		touchscreen-size-y = <2340>;
+		fts,regulator_dvdd = "vdd";
+		fts,regulator_avdd = "avdd";
+
+		fts,irq-gpio = <&tlmm 9 GPIO_TRANSITORY>;
+		fts,irq-flags = <0x2008>; /* oneshot and trigger low */
+
+		fts,reset-gpio-enable;
+		fts,reset-gpio = <&tlmm 8 GPIO_ACTIVE_HIGH>;
+
+		fts,x-max = <1080>;
+		fts,y-max = <2340>;
 	};
+	/* goodix,gt9886 (touchscreen) @ 5d */
 };
 
 &i2c9 {
 	clock-frequency = <100000>;
 	status = "okay";
 
-	/* asahi-kasei,akm09970 (magnetometer) @ 0c */
-	/* qcom,smb1390 (charger) @ 10 */
+	tfa9874: codec@34 {
+		compatible = "nxp,tfa9874";
+		reg = <0x34>;
+
+		reset-gpio = <&tlmm 12 GPIO_ACTIVE_HIGH>;
+		irq-gpio = <&tlmm 56 GPIO_ACTIVE_HIGH>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&spkr_reset &spkr_int>;
+
+		#sound-dai-cells = <0>;
+	};
+	/* ti,bq2597x (charger) @ 66 */
 };
 
 &ipa {
-	firmware-name = "qcom/sm7150/xiaomi/davinci/ipa_fws.mbn";
+	firmware-name = "qcom/sm7150/xiaomi/tucana/ipa_fws.mbn";
 };
 
 &lpass_tlmm {
@@ -155,40 +242,38 @@ &nfc_default {
 	pins = "gpio31", "gpio36", "gpio67";
 };
 
+&nfc_sleep {
+	pins = "gpio31", "gpio36", "gpio67";
+};
+
 &nxp {
 	enable-gpios = <&tlmm 67 GPIO_ACTIVE_HIGH>;
 };
 
 &panel {
-	compatible = "samsung,ams639rq08"; // or visionox,g1639fp106
+	compatible = "visionox,g2647fb105";
 
 	vsn-supply = <&vreg_ibb>;
 	vsp-supply = <&vreg_lab>;
 };
 
-&pm6150_gpios {
-	gpio-line-names = "NC", /* GPIO_1 */
-			  "NC",
-			  "SMB_STAT",
-			  "NC",
-			  "SLB",
-			  "NC",
-			  "WCSS_PWR_REQ",
-			  "NC",
-			  "FORCED_USB_BOOT",
-			  "NC";  /* GPIO_10 */
-};
-
-&pm6150_vib {
-	status = "okay";
-};
-
 &pm6150l_flash {
 	status = "okay";
 
-	led-2 {
+	led-0 {
 		function = LED_FUNCTION_FLASH;
-		color = <LED_COLOR_ID_YELLOW>;
+		function-enumerator = <0>;
+		color = <LED_COLOR_ID_WHITE>;
+		led-sources = <1>;
+		led-max-microamp = <150000>;
+		flash-max-microamp = <1000000>;
+		flash-max-timeout-us = <1280000>;
+	};
+
+	led-1 {
+		function = LED_FUNCTION_FLASH;
+		function-enumerator = <1>;
+		color = <LED_COLOR_ID_WHITE>;
 		led-sources = <2>;
 		led-max-microamp = <150000>;
 		flash-max-microamp = <1000000>;
@@ -196,53 +281,68 @@ led-2 {
 	};
 };
 
+&pm6150_gpios {
+	gpio-line-names = "FOD_LDO_EN1", /* GPIO_1 */
+			  "FOD_LDO_EN2",
+			  "NC",
+			  "VBUS_NTC_CTRL",
+			  "SLB",
+			  "NC",
+			  "NC",
+			  "NC",
+			  "FORCED_USB_BOOT",
+			  "NC";  /* GPIO_10 */
+};
+
 &pm6150l_gpios {
 	gpio-line-names = "NC", /* GPIO_1 */
 			  "KYPD_VOLP_N",
 			  "CAMW_LDO_EN",
-			  "NC",
+			  "SF_DRV_STR",
 			  "NC",
 			  "NC",
 			  "SLB",
-			  "CAMT_LDO_EN",
+			  "CAMT1_LDO_EN",
 			  "LCD_RESET_N",
-			  "MOTOR_STEP_CLK", /* GPIO_10 */
-			  "CAMT_VCM_2P8_EN",
-			  "NC";
+			  "PWM_HAPTICS", /* GPIO_10 */
+			  "CAMT2_LDO_EN",
+			  "SF_DRV_HWEN";
 };
 
-&pm6150l_lpg {
-	status = "okay";
-
-	led@1 {
-		status = "disabled";
-	};
-
-	led@2 {
-		reg = <2>;
-		color = <LED_COLOR_ID_RED>;
-		function = LED_FUNCTION_STATUS;
-		function-enumerator = <0>;
+&q6afedai {
+	dai@16 {
+		reg = <PRIMARY_MI2S_RX>;
+		qcom,sd-lines = <0>;
 	};
+};
 
-	led@3 {
-		reg = <3>;
-		color = <LED_COLOR_ID_RED>;
-		function = LED_FUNCTION_STATUS;
-		function-enumerator = <1>;
+&q6asmdai {
+	dai@0 {
+		reg = <0>;
 	};
 };
 
 &remoteproc_adsp {
-	firmware-name = "qcom/sm7150/xiaomi/davinci/adsp.mbn";
+	firmware-name = "qcom/sm7150/xiaomi/tucana/adsp.mbn";
 };
 
 &remoteproc_cdsp {
-	firmware-name = "qcom/sm7150/xiaomi/davinci/cdsp.mbn";
+	firmware-name = "qcom/sm7150/xiaomi/tucana/cdsp.mbn";
 };
 
 &remoteproc_mpss {
-	firmware-name = "qcom/sm7150/xiaomi/davinci/modem.mbn";
+	firmware-name = "qcom/sm7150/xiaomi/tucana/modem.mbn";
+};
+
+&spi8 {
+	status = "okay";
+
+	irled@0 {
+		compatible = "ir-spi-led";
+		reg = <0x0>;
+
+		spi-max-frequency = <19200000>;
+	};
 };
 
 &tlmm {
@@ -250,36 +350,36 @@ &tlmm {
 			  "NFC_ESE_SPI_MOSI",
 			  "NFC_ESE_SPI_CLK",
 			  "NFC_ESE_SPI_CS_N",
-			  "NC",
-			  "NC",
+			  "HAPTICS_EN",
+			  "CAM_DOVDD_EN",
 			  "TS_I2C_SDA",
 			  "TS_I2C_SCL",
 			  "TS_RESET_N",
 			  "TS_INT_N",
 			  "MDP_VSYNC_P", /* GPIO_10 */
-			  "NC",
-			  "NFC_ENABLE_SPKR_PA_RST",
-			  "CAMF_MCLK0",
-			  "CAMU_MCLK1",
-			  "CAMW_MCLK2",
-			  "CAMT_MCLK3",
+			  "CAMT1_RSTN",
+			  "SPKR_PA_RST",
+			  "CAM_MCLK0",
+			  "CAM_MCLK1",
+			  "CAM_MCLK2",
+			  "CAM_MCLK3",
 			  "CCI_I2C_SDA0",
 			  "CCI_I2C_SCL0",
 			  "CCI_I2C_SDA1",
 			  "CCI_I2C_SCL1", /* GPIO_20 */
-			  "NC",
+			  "CAMM_LDO_EN",
 			  "FL_STROBE_TRIG",
 			  "CAMF_RSTN",
-			  "CAMT_RSTN",
+			  "CAMM_RSTN",
 			  "CAMU_RSTN",
 			  "CAMW_RSTN",
 			  "CCI_I2C_SDA2",
 			  "CCI_I2C_SCL2",
 			  "CAMU_LDO_EN",
-			  "NC",  /* GPIO_30 */
+			  "CAMT2_RSTN",  /* GPIO_30 */
 			  "NFC_LABBCLK3_EN",
 			  "ERR_INT_N",
-			  "AUDIO_SWITCH_EN",
+			  "LCD_ID_DET1",
 			  "NFC_I2C_SDA",
 			  "NFC_I2C_SCL",
 			  "NFC_DWL_REQ",
@@ -288,8 +388,8 @@ &tlmm {
 			  "BT_UART_RFR",
 			  "BT_UART_TX", /* GPIO_40 */
 			  "BT_UART_RX",
-			  "LCD_ID_DET1",
-			  "MOTOR_FAULT",
+			  "NC",
+			  "INFARED_SPI_MOSI",
 			  "BLSP2_UART_TX",
 			  "BLSP2_UART_RX",
 			  "APPS_I2C_SDA",
@@ -299,26 +399,26 @@ &tlmm {
 			  "SPKR_I2S_WS", /* GPIO_50 */
 			  "SPKR_I2S_DOUT",
 			  "SPKR_I2S_DIN1",
-			  "SPKR_I2C_SDA",
-			  "SPKR_I2C_SCL",
-			  "MOTOR_DIR",
+			  "TYPC_I2C_SDA",
+			  "TYPC_I2C_SCL",
+			  "LCD_ID_DET2",
 			  "SPKR_INT",
 			  "CAMF_LDO_EN",
-			  "NC",
+			  "BQ25970_MASTER_INT",
 			  "FOD_SPI_MISO",
 			  "FOD_SPI_MOSI", /* GPIO_60 */
 			  "FOD_SPI_SCLK",
 			  "FOD_SPI_CS_N",
 			  "FOD_LDO_EN",
-			  "DIGITAL_HALL1_RSTN",
+			  "GPOUT",
 			  "FOD_RST",
-			  "MOTOR_M1",
-			  "SPKR_PA_RST_NFC_ENABLE",
-			  "TS_TA",
+			  "SF_DRV_TX",
+			  "NFC_ENABLE",
+			  "TOF_INT",
 			  "FOD_INT",
 			  "RFFE6_CLK", /* GPIO_70 */
 			  "RFFE6_DATA",
-			  "FM_GPIO",
+			  "TP_VDDI_EN",
 			  "WLAN_SW_CTRL",
 			  "GRFC_4",
 			  "UIM2_DATA",
@@ -334,38 +434,59 @@ &tlmm {
 			  "WMSS_RESET_N",
 			  "ACCEL_INT",
 			  "GYRO_INT",
-			  "MOTOR_SLEEP",
+			  "MIPI_SEL1",
 			  "ALS_INT_N",
-			  "MOTOR_EN", /* GPIO_90 */
-			  "MOTOR_M0",
+			  "MIPI_SEL2", /* GPIO_90 */
+			  "SF_DRV_TORCH",
 			  "RFFE4_CLK",
-			  "DIGITAL_HALL1_INT",
 			  "NC",
+			  "SW_IO",
 			  "GRFC_3",
 			  "QLINK_REQUEST",
 			  "QLINK_ENABLE",
 			  "GRFC_2",
 			  "NC",
-			  "GRFC_1", /* GPIO_100 */
+			  "NC", /* GPIO_100 */
 			  "RFFE3_DATA",
 			  "RFFE3_CLK",
 			  "RFFE4_DATA",
 			  "USB_PHY_PS",
 			  "RFFE5_DATA",
 			  "RFFE5_CLK",
-			  "NC",
+			  "GRFC_5",
 			  "WLAN_COEX_UART_TXD",
 			  "WLAN_COEX_UART_RXD",
-			  "NC", /* GPIO_110 */
-			  "NC",
+			  "GRFC_34", /* GPIO_110 */
+			  "GRFC_33",
 			  "RFFE1_DATA",
 			  "RFFE1_CLK";
+
+	spkr_reset: spkr-reset-state {
+		pins = "gpio12";
+		function = "gpio";
+		drive-strength = <2>;
+		bias-disable;
+		bias-pull-down;
+		output-low;
+	};
+
+	spkr_int: spkr-int-state {
+		pins = "gpio56";
+		function = "gpio";
+		drive-strength = <2>;
+		bias-pull-up;
+		input-enable;
+	};
+};
+
+&uart8 {
+	status = "disabled";
 };
 
 &venus {
-	firmware-name = "qcom/sm7150/xiaomi/davinci/venus.mbn";
+	firmware-name = "qcom/sm7150/xiaomi/tucana/venus.mbn";
 };
 
 &wifi {
-	qcom,ath10k-calibration-variant = "xiaomi_davinci";
+	qcom,ath10k-calibration-variant = "xiaomi_tucana";
 };
